syntax = "proto3";

package message;


enum feature_t {
    none = 0;
    bats = 1;
    sb = 2;
    min = 3;
    maj = 4;
    capital = 5;
}

message hex_coord_t
{
    int32 x = 1;
    int32 y = 2;
}

message capital_hex_zone_t
{
    string name = 1;
    repeated feature_t features = 2 [packed=true];
}

message capital_hex_t
{
    int32 coord = 1;
    repeated capital_hex_zone_t zones = 2;
}

message capital_t
{
    repeated capital_hex_t hexes = 1;
}

message offmap_t
{
    int32 provinces = 1;
    int32 mins = 2;
    int32 majs = 3;
}

message nation_t
{
    string name = 1;
    string short_name = 2;
    capital_t capital = 3;
    int32 free_strategic_moves = 4;
    int32 capital_star_points = 5;
    offmap_t offmap = 6;
    int32 offmap_survey_ships = 7;
    int32 nation_id = 8;
    int32 cannot_build_offmap_capital = 9;
    int32 old_offmap_shipyard = 10;
}

message nations_t
{
    map<string, nation_t> nations = 1;
}

message hex_t
{
    hex_coord_t coord = 1;
    uint32 owner = 2;
    feature_t feature = 3;
    repeated int32 neutral_zone_bordering = 4;
}

message province_t
{
    repeated int32 hexes = 1;
}

message province_list_t
{
    repeated province_t provinces = 1;
}

message map_t
{
    int32 width = 1;
    int32 height = 2;
    map<int32, province_list_t> provinces = 3;
    repeated hex_t hexes = 4;
}

message model_t
{
    nations_t nations = 1;
    map_t map = 2;
}
